File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.py
"""
모든 경우의 수를 다 비교하자

1. discount의 경우의 수를 모두 구한다.
2. 각 경우에서 이모티콘 플러스 서비스 가입자와 판매액을 구한다.
"""

from itertools import product
from dataclasses import dataclass

@dataclass
class Emoticon:
    price: int
    discount: int
    
    @property
    def discounted_price(self):
        return self.price * (100 - self.discount) // 100

def solution(users, emoticons):
    emoticon_cases = [[Emoticon(p, d) for p, d in zip(emoticons, discount)] for discount in product([10, 20, 30, 40], repeat=len(emoticons))]
    
    max_plus = 0
    max_total = 0
    for emoticon_case in emoticon_cases:
        plus = 0
        total = 0
        for discount, threshold in users:
            user_total = 0
            for emoticon in emoticon_case:
                if emoticon.discount < discount:
                    continue
                user_total += emoticon.discounted_price
            if user_total >= threshold:
                plus += 1
            else:
                total += user_total
        if max_plus < plus:
            max_plus = plus
            max_total = total
        elif max_plus == plus:
            max_total = max(max_total, total)
        
    return [max_plus, max_total]

File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.java
import java.util.*;
import java.util.stream.Collectors;

class Solution {

    static class Emoticon {
        int price;
        int discount;

        Emoticon(int price, int discount) {
            this.price = price;
            this.discount = discount;
        }

        int getDiscountedPrice() {
            return price * (100 - discount) / 100;
        }
    }

    private static final int[] PERCENTS = {10, 20, 30, 40};

    public int[] solution(int[][] users, int[] emoticons) {
        int n = emoticons.length;
        List<List<Integer>> discountCases = generateDiscountCases(n);

        int maxPlus = 0;
        int maxTotal = 0;

        for (List<Integer> discountCase : discountCases) {
            List<Emoticon> emoticonCase = new ArrayList<>();
            
            for (int i = 0; i < n; i++) {
                emoticonCase.add(new Emoticon(emoticons[i], discountCase.get(i)));
            }
            
            int plus = 0;
            int total = 0;
            
            for (int[] user : users) {
                int userTotal = 0;
                int minDiscount = user[0];
                int thresholdTotal = user[1];
                
                for (Emoticon emoticon : emoticonCase) {
                    if (emoticon.discount >= minDiscount) userTotal += emoticon.getDiscountedPrice();
                }
                
                if (userTotal >= thresholdTotal) {
                    plus += 1;
                    continue;
                } else {
                    total += userTotal;
                }
            }
            
            if (plus > maxPlus) {
                maxPlus = plus;
                maxTotal = total;
            } else if (plus == maxPlus) {
                maxTotal = Math.max(maxTotal, total);
            }
        }
        
        return new int[]{maxPlus, maxTotal};
    }

    private static List<List<Integer>> generateDiscountCases(int n) {
        List<List<Integer>> discountCases = Arrays.stream(PERCENTS)
                .boxed()
                .map(List::of)
                .collect(Collectors.toList());

        for (int i = 1; i < n; i++) {
            List<List<Integer>> currentCases = new ArrayList<>();

            for (List<Integer> discountCase : discountCases) {
                for (int p : PERCENTS) {
                    List<Integer> copiedCase = new ArrayList<>(discountCase);
                    copiedCase.add(p);
                    currentCases.add(copiedCase);
                }
            }
            discountCases = currentCases;
        }
        return discountCases;
    }
}